gateway-js/src/index.ts

```diff
@@ -472,6 +520,108 @@ export class ApolloGateway implements GraphQLService {
     }
   }
 
+  private async updateWithCsdl(result: CsdlUpdate): Promise<void> {
+    if (result.id === this.compositionId) {
+      this.logger.debug('No change in composition since last check.');
+      return;
+    }
+
+    // This may throw, so we'll calculate early (specifically before making any updates)
+    // In the case that it throws, the gateway will:
+    //   * on initial load, throw the error
+    //   * on update, log the error and don't update
+    const parsedCsdl = parse(result.csdl);
+
+    const previousSchema = this.schema;
+    const previousCsdl = this.parsedCsdl;
+    const previousCompositionId = this.compositionId;
+
+    if (previousSchema) {
+      this.logger.info('Updated CSDL was found.');
+    }
+
+    await this.maybePerformServiceHealthCheck(result);
+
+    this.compositionId = result.id;
+    this.parsedCsdl = parsedCsdl;
+
+    if (this.queryPlanStore) this.queryPlanStore.flush();
+
+    const { schema, composedSdl } = this.createSchemaFromCsdl(result.csdl);
+
+    if (!composedSdl) {
+      this.logger.error(
+        "A valid schema couldn't be composed. Falling back to previous schema.",
+      );
+    } else {
+      this.schema = schema;
+      this.queryPlannerPointer = getQueryPlanner(composedSdl);
+
+      // Notify the schema listeners of the updated schema
+      try {
+        this.onSchemaChangeListeners.forEach((listener) =>
+          listener(this.schema!),
+        );
+      } catch (e) {
+        this.logger.error(
+          "An error was thrown from an 'onSchemaChange' listener. " +
+            'The schema will still update: ' +
+            ((e && e.message) || e),
+        );
+      }
+
+      if (this.experimental_didUpdateComposition) {
+        this.experimental_didUpdateComposition(
+          {
+            compositionId: result.id,
+            csdl: result.csdl,
+            schema: this.schema,
+          },
+          previousCompositionId && previousCsdl && previousSchema
+            ? {
+                compositionId: previousCompositionId,
+                csdl: print(previousCsdl),
+                schema: previousSchema,
+              }
+            : undefined,
+        );
+      }
+    }
+  }
+
+  private async maybePerformServiceHealthCheck(update: CompositionUpdate) {
+    // Run service health checks before we commit and update the new schema.
+    // This is the last chance to bail out of a schema update.
+    if (this.config.serviceHealthCheck) {
+      const serviceList = isCsdlUpdate(update)
+        ? // Parsing could technically fail and throw here, but parseability has
+          // already been confirmed slightly earlier in the code path
+          this.serviceListFromCsdl(parse(update.csdl))
```

Comment:
Can we comment-and-link-to-issue that this is a redundant parse?

Reply:
80577a3